parser.py

```diff
@@ -0,0 +1,168 @@
+import scanner
+import ply.yacc as yacc
+import lexer as scanner
+
+
+tokens = scanner.tokens
+
+precedence = (
+   ("nonassoc", 'IF'),
+   ("nonassoc", 'ELSE'),
+   ("nonassoc", '<', '>', 'EQ', 'NOTEQ', 'GE', 'LE'),
+   ("left", '+', '-'),
+   ("left", 'DOTADD', 'DOTSUB'),
+   ("left", '*', '/'),
+   ("left", 'DOTMUL', 'DOTDIV'),
+   ("left", 'TRANSPOSE'),
+   ("right", 'UMINUS')
+)
+
+
+def p_error(p):
+    if p:
+        print("Syntax error at line {0}: LexToken({1}, '{2}')".format(p.lineno, p.type, p.value))
+    else:
+        print("Unexpected end of input")
+
+
+def p_program(p):
+    """program : instructions_opt"""
+
+def p_instructions_opt_1(p):
+    """instructions_opt : instructions """
+
+def p_instructions_opt_2(p):
+    """instructions_opt : """
+
+def p_instructions_1(p):
+    """instructions : instructions instruction """
+
+def p_instructions_2(p):
+    """instructions : instruction """
+
+def p_instruction(p):
+    """ instruction :  if
+                    |  for
+                    |  while
+                    |  break
+                    |  continue
+                    |  print
+                    |  block
+                    |  return 
+                    |  assign 
+                    |  assign_array
+                    """
+
+def p_num(p):
+    """ num : INTNUM
+            | FLOAT """
+
+def p_val(p):
+    """ value : STRING
+              | num """
+
+def p_val_array(p):
+    """ values : value
+               | value ',' values"""
+
+def p_rows(p): #TODO
+    """ rows : '[' values ']'
+             | '[' values ']' ',' rows """
+
+def p_array(p): #TODO
+    """ array : '[' ']'
+              | '[' rows ']'"""
+
+def p_array_val(p):
+    """ array_val : '[' INTNUM ',' INTNUM ']' 
+                  | '[' INTNUM ']' """
+
+def p_expression(p):
+    """ expression : num
+                   | ID """
+
+def p_expression_transpose(p):
+    """ expression : ID TRANSPOSE """
+
+def p_expression_uminus(p):
+    """ expression : '-' expression %prec UMINUS """
+
+def p_expression_bin_op(p):
+    """ expression : expression '+' expression 
+                   | expression '-' expression
+                   | expression '*' expression
+                   | expression '/' expression
+                   | expression DOTADD expression
+                   | expression DOTSUB expression
+                   | expression DOTMUL expression
+                   | expression DOTDIV expression"""
```

Comment:
hmm, expression może być rozszerzone do, `num`, `ID`, `ID TRANSPOSE`, `-expression` jeżeli rozumiem dobrze,
czyli terminale tutaj to jest `num`, `ID` i `ID TRANSPOSE`, czyli `expression DOTADD expression` może wygenerować w najlepszym razie `DOTADD` miedzy elementami które nawet nie są macierzami... dobrze piszę czy coś źle rozumiem?